### PR TITLE
Feat/add cloud stream utils

### PR DESCRIPTION
**Pull request by dominikhorn93**
---
name: Pull request
about: Contribute your changes to the project
title: '[pull-request]'
labels: ''
assignees: ''
---
**Description**
Add spring cloud stream utils and start process endpoint. 

**Reference**
Issues [#XXX](https://github.com/it-at-m/digiwf-project/issues/298)

**Checklist**

- [x] Acceptance criteria are met
- [x] Pipeline has been run successfully
- [x] JUnit tests have been run successfully
- [x] Changelog is adjusted
- [x] [Documentations](https://git.muenchen.de/groups/digitalisierung/-/wikis/Dokumentationen) are completed
- [ ] Frontend is tested
- [ ] Created sub-branches are deleted
- [ ] Boards are updated ( [Digitalisierung](https://git.muenchen.de/groups/digitalisierung/-/boards) , [Support](https://git.muenchen.de/digitalisierung/digiwf-support) , [Zenhub](https://app.zenhub.com/workspaces/digiwf-621f70bf50ea1100120b7e93/board) )
- [ ] feature process is created or extended
- [ ] [git-ops](https://git.muenchen.de/digitalisierung/digiwf-ops) is customized
- [ ] Openshift environments are prepared (Secrets, etc.)

### CODE REVIEW

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/pom.xml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    <artifactId>digiwf-spring-cloudstream-utils-core</artifactId>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework.cloud</groupId>
+            <artifactId>spring-cloud-stream</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.muenchendigital.digiwf</groupId>
```
Comment (martind260) [inline digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/pom.xml:20]: @dominikhorn93 Is this dependency needed?

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/bpmnerror/service/BpmnErrorService.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+        final Message<BpmnErrorEvent> message = MessageBuilder
+                .withPayload(builder.build())
+                .setHeader(StreamingHeaders.TYPE, MESSAGE_TYPE)
+                .build();
+
+        final Sinks.EmitResult emitResult = this.bpmnErrorSink.tryEmitNext(message);
+
+        if (emitResult.isSuccess()) {
+            log.info("The error {} was successfully delivered to the eventbus.", message.getHeaders().get(MessageHeaders.ID));
```
Comment (martind260) [inline digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/bpmnerror/service/BpmnErrorService.java:45]: @dominikhorn93 Wrong message?

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/message/service/CorrelateMessageService.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Sinks;
+
+import java.util.Map;
+
+import static io.muenchendigital.digiwf.spring.cloudstream.utils.api.streaming.infrastructure.StreamingHeaders.DIGIWF_MESSAGE_NAME;
+import static io.muenchendigital.digiwf.spring.cloudstream.utils.api.streaming.infrastructure.StreamingHeaders.DIGIWF_PROCESS_INSTANCE_ID;
+
+@Slf4j
```
Comment (martind260) [inline digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/message/service/CorrelateMessageService.java:21]: @dominikhorn93 What about javadoc?

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import reactor.core.publisher.Sinks;
+
+import java.util.Map;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class StartProcessService {
+
+    private static final String CORRELATEMESSAGEV_01 = "startProcessV01";
```
Comment (martind260) [inline digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java]: @dominikhorn93 Adjust constant name.

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                .withPayload(startProcessEvent)
+                .setHeader(StreamingHeaders.TYPE, CORRELATEMESSAGEV_01)
+                .build();
+
+//        // TODO: If spring.cloud.function.definition=sendCorrelateMessage is not set, messaging doesnt work, but no error is thrown.
+//        // Same for spring.cloud.stream.bindings.sendCorrelateMessage-out-0.destination=
+        final Sinks.EmitResult emitResult = this.startProcessSink.tryEmitNext(message);
+//
+        if (emitResult.isSuccess()) {
+            log.debug("The correlating message {} was successfully delivered to the eventbus.", message.getHeaders().get(MessageHeaders.ID));
```
Comment (martind260) [inline digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java]: @dominikhorn93 Check log message

**Comment by martind260** [comment]
@dominikhorn93 Will you provide your dependency image as documenation?

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/pom.xml (hunk 1 above)
Comment (martind260) [reply digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/pom.xml:20]: Shouldnt it be api?

Comment (dominikhorn93) [reply digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/pom.xml:20]: Yes!

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/message/service/CorrelateMessageService.java (hunk 3 above)
Comment (dominikhorn93) [reply digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/message/service/CorrelateMessageService.java:21]: Added 👍

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java (hunk 4 above)
Comment (dominikhorn93) [reply digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java]: Changed 👍

File: digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java (hunk 5 above)
Comment (dominikhorn93) [reply digiwf-libs/digiwf-spring-cloudstream-utils/digiwf-spring-cloudstream-utils-core/src/main/java/io/muenchendigital/digiwf/spring/cloudstream/utils/api/streaming/process/service/StartProcessService.java]: Thanks!

**Comment by dominikhorn93** [comment]
> @dominikhorn93 Will you provide your dependency image as documenation?

Sure - Right now that only covers the digiwf-connector module. I think we should add the documentation to the digiwf-project?

**Review by martind260 — APPROVED** [review]

**Pull request merged**